### PR TITLE
fixed UX problem in the state selector in customers admin

### LINKED ISSUE #2106: When I update customer address, confusing UX on states when changing country
## Description
In customer address I see Australia, choose Australian state. Then change country to France, I still see the Australian state. I click to change the state and see the list of French states so I can choose the right state, but this period of seeing one country associated with states from another country is pretty confusing. 

## Expected Behavior
If I change country in customer address edit form, states field empty dynamically before I choose the right state from the new states list proposed.

## Actual Behavior
If I change country in customer address edit form, states field remain with the previous state chosen. I can then choose the right state from the new states list proposed, but it's confusing.

## Steps to Reproduce
1. Make sure you use an instance with states for at least two countries
2. Go to customer, select a shop, click to edit one customer address, and try to change countries and states.

## Animated Gif/Screenshot
https://www.useloom.com/share/f20798787a9d48abb5f7eeebf454159f 

## Context
Remaining small bug from PR #2074 that got merged (as was fixing the main issue).

## Severity
S4

## Your Environment
- Version used: Aus staging with branch PR2074
- Browser name and version: Version 64.0.3282.167 (Official Build) (64-bit)
- Operating System and version (desktop or mobile): Ubuntu 16.04 LTS

### PR DESCRIPTION
**Pull request by luisramos0**
#### What? Why?

Closes #2106

We needed to clear the state selector when a new country is selected.
The list of states was being updated but the states select box was not being cleared. Code to clear the selector was added whenever the country select is updated.

#### What should we test?

Verify that the issue is solved.

#### Release notes

Fixed UX problem in the state selector in customers admin.

### CODE REVIEW

File: app/assets/javascripts/admin/customers/directives/edit_address_dialog.js.coffee [hunk 1]
```diff
@@ -25,13 +26,18 @@ angular.module("admin.customers").directive 'editAddressDialog', ($compile, $tem
...
       scope.$apply()
 
-    scope.filter_states = (countryID) ->
+    scope.filterStates = (countryID) ->
       return [] unless countryID
-      $filter('filter')(scope.availableCountries, {id: parseInt(countryID)}, true)[0].states
\ No newline at end of file
+      $filter('filter')(scope.availableCountries, {id: parseInt(countryID)}, true)[0].states
+
+    scope.clearInvalidState = ->
```
Comment (sauloperez) [inline app/assets/javascripts/admin/customers/directives/edit_address_dialog.js.coffee]: I think a name like `clearStateOfAnotherCountry` or something similar would better convey what this function is meant to do. Did I understand it correctly?

Comment (oeoeaio) [reply app/assets/javascripts/admin/customers/directives/edit_address_dialog.js.coffee]: I am a splitter, so I would do:
```coffee
scope.clearState() unless scope.addressStateMatchesCountry()

...

scope.clearState = ->
  scope.address.state_id = ""

scope.addressStateMatchesCountry = ->
  scope.states.some (state) -> state.id == scope.address.state_id
```

Comment (sauloperez) [reply app/assets/javascripts/admin/customers/directives/edit_address_dialog.js.coffee]: That's even better! +1

Comment (luisramos0) [reply app/assets/javascripts/admin/customers/directives/edit_address_dialog.js.coffee]: all right, looks great. thanks a lot guys! done.

**Comment by luisramos0** [comment]
Hey @sauloperez 
I didn't manage to create ng tests for directive edit-address-dialog, so I added the test of the bug reported above to customers_spec.rb. Is this ok? What do you think?

**Comment by sauloperez** [comment]
I'm starting to worry about us relying solely on integration testing. I feel like this is something we need to change soon. I can proof with it with data but I believe we ended up with an upside-down testing pyramid 😅 .

@oeoeaio do you have any idea on how to add unit tests for this. Angular testing is not my area of expertise.

**Comment by luisramos0** [comment]
I agree. And this is a good example where it should be done at unit test level (edit address dialog). I just cant make the directive test work...

This is what I have testing edit_address_dialog.js:
https://gist.github.com/luisramos0/9139da2ae89cf87e5063b0328c374555
I don't know how to initialise the directive with injected dependencies like in the line starting with #$controller

Cheers,
Luís

**Comment by oeoeaio** [comment]
@luisramos0 you should be able to achieve this using ngMock's `module` function by injecting `$provide`, you can use it like this immediately below your existing module registration code:

```coffee
  beforeEach ->
    module 'admin.customers'
    module ($provide) ->
      $provide.value 'dependencyNameHere', 'some value for the dependency'
      null
```

Make sure that you return `null` from the module call, I can't remember what exactly, but I seem to remember that bad things happen when coffeescript implicitly returns the result of a `$provide.value` call... 🤦‍♀️

**Comment by myriamboure** [comment]
Ok so who would like to take that over? @frank-west-iii do you think that's something you could work on? Doesn't seem easy apparently... let me know what you think.

**Comment by luisramos0** [comment]
To make it clear: the problem is fixed and tested with a functional test, the challenge is to test this at directive level. What I am having trouble with is setting up a directive test (there are none in the project). Testing the change itself should be straight forward.

**Comment by myriamboure** [comment]
Ah, ok @luisramos0 ... it's just the tests I understand ... and I guess it's good not to spend too much time also if you block on something... @oeoeaio @sauloperez any idea who could help on that? I don't see the technicality of it an don't know all the dev tech skills...

**Comment by sauloperez** [comment]
@oeoeaio can you give it a try? looks like you did this before.

**Comment by oeoeaio** [comment]
I didn't realise how huge this was. I have just to write a test and utterly failed. In its current state, I think it is completely untestable. I've done a bit of a spike to refactoring this into three separate, more testable components (progress [here](https://github.com/oeoeaio/openfoodnetwork/commit/fbb21a36da3c5403674a5f14744841978eb00e47)), but I've already spent far too long on it. I will try and come back to it when I have more time, or someone else can take over from where I got up to.

**Comment by luisramos0** [comment]
That refactor looks great (it's great to learn how things should be structured :-))
I can pick it up and fix anything that could need fixing in your refactor.

But I am still wondering how do I test the error case I wanted (on the second time you open the dialog, make sure that the dialog doesn't show errors from the previous interaction). It needs a test with directive and template...

**Comment by sauloperez** [comment]
if would rather create an issue if we want to go that way @luisramos0 but I wouldn't excessive time on this.

**Comment by oeoeaio** [comment]
> But I am still wondering how do I test the error case I wanted

@luisramos0 for that case, is it enough to just check that the `errors` array on the scope is cleared? There is not really any need to test the binding between the scope and the template is working, that is Angular's domain.

**Comment by oeoeaio** [comment]
Or do you just mean that you need the directive to simulate the opening of the modal? Hm... 🤔

**Comment by luisramos0** [comment]
Hi @oeoeaio testing the errors array is still in the dialog directive. And I didnt manage to setup a directive test with the dependencies (with or without the template).
I think your refactor is great and we should pick it up at some point. This may take a while and so I am skipping it for now.
Regarding the test, if you can setup a test of this directive with the dependencies and can pick that up and test the errors array, otherwise I don't think we should spend more time around this.
What I'd do with this:
* if @oeoeaio can setup directive test with dependencies, write the directive test
  ** otherwise, let's ignore/skip this, there's a functional test covering this
  ** @oeoeaio, can you please see if you want to skip this and go ahead with this change as is, or if you you can provide a test of the directive (with the version on my PR without your refactor)?
* anyway, I'd create an issue to refactor edit-address-dialog and pick up @oeoeaio  refactoring commit

I hope this makes sense ;-)

**Comment by daniellemoorhead** [comment]
OK, as much as there has been a lot of good and time consuming work done on this it's now almost 2 months since it was touched...and so I'm going to close it and put the issue back to dev ready.

**Comment by luisramos0** [comment]
i think we should move forward instead.
the problem is fixed and automaticaly tested with a functional test, the challenge is to test this at directive/lower level.
Pau did require a test at directive level which is something we dont have on ofn and I didnt manage to make it work.
so id challenge Pau to accept the functional test as enough and move forward.

**Comment by Matt-Yorkley** [comment]
Lots of work has gone into this so it probably shouldn't be wasted. The discussions above were related to lack of angular unit tests, but we can open a separate PR if that is needed. The code looks good, it's covered by a functional test, and the previous request for changes from Pau was later changed to an approval, so I think we can move to `test ready`.

**Comment by daniellemoorhead** [comment]
Great, glad to see it being picked up again. Let's get it through to done!

**Comment by daniellemoorhead** [comment]
semaphore is failing...needs to a rebase @luisramos0 ?

**Comment by luisramos0** [comment]
yes, rebase fixed it 👍 
ready for staging now.

**Comment by Matt-Yorkley** [comment]
Staged on: https://staging.openfoodnetwork.org.uk

**Comment by RachL** [comment]
This is good to go !

Testing notes : https://docs.google.com/document/d/1_L_7S4kYb9lzXz3bPepijo7rPcL2vU_fUw33ADnK-wI/edit#

Thanks @myriamboure for the video in original issue :)

**Pull request merged**